src/front/wgsl/mod.rs

```diff
@@ -2045,104 +2378,222 @@ impl Parser {
             self.scopes.last().map(|&(ref scope, _)| scope.clone()),
             Some(Scope::PrimaryExpr)
         );
-        let ty_resolution = match self.lookup_type.get(type_name) {
-            Some(&handle) => TypeResolution::Handle(handle),
-            None => match self.parse_type_decl_impl(
-                lexer,
-                TypeAttributes::default(),
-                type_name,
-                ctx.types,
-                ctx.constants,
-            )? {
-                Some(inner) => TypeResolution::Value(inner),
-                None => return Ok(None),
-            },
+        let dst_ty = match self.lookup_type.get(type_name) {
+            Some(&handle) => ConstructorType::Struct(handle),
+            None => {
+                match self.parse_constructor_type(lexer, type_name, ctx.types, ctx.constants)? {
+                    Some(inner) => inner,
+                    None => {
+                        match self.parse_type_decl_impl(
+                            lexer,
+                            TypeAttributes::default(),
+                            type_name,
+                            ctx.types,
+                            ctx.constants,
+                        )? {
+                            Some(_) => {
+                                let span = self.pop_scope(lexer);
+                                return Err(Error::TypeNotConstructible(span));
+                            }
+                            None => return Ok(None),
+                        }
+                    }
+                }
+            }
         };
 
+        lexer.open_arguments()?;
+
         let mut components = Vec::new();
-        let (last_component, arguments_span) = lexer.capture_span(|lexer| {
-            lexer.open_arguments()?;
-            let mut last_component = self.parse_general_expression(lexer, ctx.reborrow())?;
+        let mut spans = Vec::new();
+
+        if lexer.peek().0 == Token::Paren(')') {
+            let _ = lexer.next();
+        } else {
+            while components.is_empty() || lexer.next_argument()? {
+                let (component, span) = lexer
+                    .capture_span(|lexer| self.parse_general_expression(lexer, ctx.reborrow()))?;
+                components.push(component);
+                spans.push(span);
+            }
+        }
+
+        enum Components<'a> {
+            None,
+            One(Handle<crate::Expression>, Span, &'a crate::TypeInner),
+            Many(
+                Vec<Handle<crate::Expression>>,
+                Vec<Span>,
+                // type of the first component
+                &'a crate::TypeInner,
+            ),
+        }
 
-            while lexer.next_argument()? {
-                components.push(last_component);
-                last_component = self.parse_general_expression(lexer, ctx.reborrow())?;
+        impl<'a> Components<'a> {
+            fn components(self) -> Vec<Handle<crate::Expression>> {
```

Comment:
maybe `to_components_vec`? The `components()` method looks like an accessor, but it's not an accessor because it creates a vector and returns it.

Reply:
How about `into_components_vec` (since it takes ownership)?

Reply:
oh yes, that sounds great!